flagsmith-engine/features/models.ts

```diff
@@ -0,0 +1,133 @@
+
+    constructor(
+        multivariate_feature_option: MultivariateFeatureOptionModel,
+        percentage_allocation: number,
+        id: number,
+        mvFsValueUuid?: string
+    ) {
+        this.id = id;
+        this.percentageAllocation = percentage_allocation;
+        this.multivariateFeatureOption = multivariate_feature_option;
+        this.mvFsValueUuid = mvFsValueUuid || this.mvFsValueUuid;
+    }
+}
+
+export class FeatureStateModel {
+    feature: FeatureModel;
+    enabled: boolean;
+    djangoID: number;
+    featurestateUUID: string = uuidv4();
+    featureSegment?: FeatureSegment;
+    private value: any;
+    multivariateFeatureStateValues: MultivariateFeatureStateValueModel[] = [];
+
+    constructor(
+        feature: FeatureModel,
+        enabled: boolean,
+        djangoID: number,
+        value?: any,
+        featurestateUuid: string = uuidv4()
+    ) {
+        this.feature = feature;
+        this.enabled = enabled;
+        this.djangoID = djangoID;
+        this.value = value;
+        this.featurestateUUID = featurestateUuid;
+    }
+
+    setValue(value: any) {
+        this.value = value;
+    }
+
+    getValue(identityId?: number | string) {
+        if (!!identityId && this.multivariateFeatureStateValues.length > 0) {
+            return this.getMultivariateValue(identityId);
+        }
+        return this.value;
+    }
+
+    /* 
+        Returns `True` if `this` is higher segment priority than `other`
+        (i.e. has lower value for featureSegment.priority)
+        NOTE:
+            A segment will be considered higher priority only if:
+            1. `other` does not have a feature segment(i.e: it is an environment feature state or it's a
+            feature state with feature segment but from an old document that does not have `featureSegment.priority`)
+            but `this` does.
+            2. `other` have a feature segment with high priority
+    */
+    isHigherSegmentPriority(other: FeatureStateModel): boolean {
+        if (!other.featureSegment || !this.featureSegment) {
+            return false;
+        }
+        return this.featureSegment.priority < other.featureSegment.priority;
+    }
+
+    getMultivariateValue(identityID: number | string) {
+        const percentageValue = getHashedPercentateForObjIds([
+            this.djangoID || this.featurestateUUID,
+            identityID
+        ]);
+
+        let startPercentage = 0;
+        const sortedF = this.multivariateFeatureStateValues.sort((a, b) =>{
+            return a.id - b.id;
+        });
+        for (const myValue of sortedF) {
+            const limit = myValue.percentageAllocation + startPercentage;
+            if (startPercentage <= percentageValue && percentageValue < limit) {
+                return myValue.multivariateFeatureOption.value;
+            }
+            startPercentage = limit;
+        }
+        return this.value;
+    }
+}
+
+export class FeatureSegment {
```

Comment:
This should only have the `priority` attribute I think? Why does it have these extra attributes? The JSON document will not include this information.